File: deploy/cloudflare/src/pricing.ts
import { canonical, isBoundedString, isCanonicalUnsignedDecimal, sha256 } from "./contracts";

const PRICE_FIELDS = ["input_e8_per_million", "output_e8_per_million", "cache_read_e8_per_million", "cache_write_e8_per_million", "cache_write_5m_e8_per_million", "cache_write_1h_e8_per_million", "image_input_e8_per_million", "image_output_e8_per_million", "priority_input_e8_per_million", "priority_output_e8_per_million", "priority_cache_read_e8_per_million", "priority_cache_write_e8_per_million"] as const;
const BPS_FIELDS = ["fast_multiplier_bps", "flex_multiplier_bps", "max_reasoning_effort_multiplier_bps"] as const;
const MAX_USAGE_DIGITS = 20;
export type PricingRule = Record<(typeof PRICE_FIELDS)[number] | (typeof BPS_FIELDS)[number], string> & { version_id: string; model_pattern: string; match_kind: "exact" | "family" };
export type AdmittedPriceCard = { version_id: string; digest: string; max_reservation_e8_usd: string; rule: PricingRule };
export type PricingSnapshot = { version_id: string; digest: string; max_reservation_e8_usd: string; rules: PricingRule[] };
export type UsageForPricing = { input_tokens: string; image_input_tokens: string; output_tokens: string; image_output_tokens: string; cache_creation_tokens: string; cache_creation_5m_tokens: string; cache_creation_1h_tokens: string; cache_read_tokens: string; service_tier: string; reasoning_effort: string; rate_multiplier_bps: string };
export type E8ChargeBreakdown = { input_e8_usd: string; image_input_e8_usd: string; output_e8_usd: string; image_output_e8_usd: string; cache_write_e8_usd: string; cache_read_e8_usd: string; total_e8_usd: string; exceeds_reservation_cap: boolean };
type VersionRow = Omit<PricingSnapshot, "rules">;
type Part = { numerator: bigint; denominator: bigint; index: number };
export class PricingUnavailableError extends Error { constructor(message: string) { super(message); this.name = "PricingUnavailableError"; } }

export function normalizePricingModel(value: unknown): string | null {
  if (!isBoundedString(value, 256)) return null;
  let normalized = value.trim().toLowerCase();
  if (!/^[a-z0-9._:/-]+$/.test(normalized)) return null;
  if (normalized.startsWith("claude-")) normalized = normalized.replaceAll(".", "-");
  return normalized;
}
export function normalizePricingPattern(value: unknown, kind: unknown): string | null {
  if ((kind !== "exact" && kind !== "family") || typeof value !== "string") return null;
  if (kind === "family" && (!value.endsWith("*") || value.slice(0, -1).includes("*"))) return null;
  const normalized = normalizePricingModel(kind === "family" ? value.slice(0, -1) : value);
  return normalized === null ? null : normalized + (kind === "family" ? "*" : "");
}
function unsigned(value: unknown, max = 40): value is string { return isCanonicalUnsignedDecimal(value) && value.length <= max; }
function positive(value: unknown, max = 40): value is string { return unsigned(value, max) && value !== "0"; }
function digest(value: unknown): value is string { return typeof value === "string" && /^[0-9a-f]{64}$/.test(value); }
function versionID(value: unknown): value is string { return typeof value === "string" && /^[a-z0-9._:-]{1,128}$/.test(value); }
function validRule(value: unknown): value is PricingRule {
  if (!value || typeof value !== "object" || Array.isArray(value)) return false;
  const rule = value as Partial<PricingRule>;
  return versionID(rule.version_id) && (rule.match_kind === "exact" || rule.match_kind === "family") &&
    normalizePricingPattern(rule.model_pattern, rule.match_kind) === rule.model_pattern &&
    PRICE_FIELDS.every((key) => unsigned(rule[key])) && BPS_FIELDS.every((key) => unsigned(rule[key], 8)) &&
    positive(rule.max_reasoning_effort_multiplier_bps, 8);
}
function overlap(a: PricingRule, b: PricingRule): boolean {
  if (a.match_kind !== "family" || b.match_kind !== "family") return false;
  const left = a.model_pattern.slice(0, -1), right = b.model_pattern.slice(0, -1);
  return left.startsWith(right) || right.startsWith(left);
}
function order(a: string, b: string): number { return a < b ? -1 : a > b ? 1 : 0; }
export async function pricingDigest(version: string, cap: string, rules: PricingRule[]): Promise<string> {
  return sha256(canonical({ version_id: version, max_reservation_e8_usd: cap, rules: [...rules].sort((a,b) => order(a.model_pattern,b.model_pattern) || order(a.match_kind,b.match_kind)) }));
}
export async function validatePricingSnapshot(version: unknown, cap: unknown, expectedDigest: unknown, rules: unknown[]): Promise<PricingSnapshot> {
  if (!versionID(version) || !positive(cap, 18) || !digest(expectedDigest) || !Array.isArray(rules) || rules.length === 0 || !rules.every(validRule) || !rules.every((rule) => rule.version_id === version)) throw new PricingUnavailableError("malformed pricing snapshot");
  const exact = new Set<string>();
  for (const rule of rules) { if (rule.match_kind === "exact") { if (exact.has(rule.model_pattern)) throw new PricingUnavailableError("ambiguous exact pricing rule"); exact.add(rule.model_pattern); } }
  for (let i=0;i<rules.length;i+=1) if (rules.slice(i+1).some((other) => overlap(rules[i],other))) throw new PricingUnavailableError("ambiguous pricing wildcard rule");
  if (await pricingDigest(version,cap,rules) !== expectedDigest) throw new PricingUnavailableError("pricing digest mismatch");
  return { version_id: version, digest: expectedDigest, max_reservation_e8_usd: cap, rules };
}
function selectRule(model: string, rules: PricingRule[]): PricingRule {
  const exact = rules.filter((rule) => rule.match_kind === "exact" && rule.model_pattern === model);
  if (exact.length === 1) return exact[0]; if (exact.length > 1) throw new PricingUnavailableError("ambiguous exact pricing rule");
  const family = rules.filter((rule) => rule.match_kind === "family" && model.startsWith(rule.model_pattern.slice(0,-1)));
  if (family.length !== 1) throw new PricingUnavailableError(family.length ? "ambiguous pricing wildcard rule" : "missing model price");
  return family[0];
}
const ruleColumns = "version_id,model_pattern,match_kind,input_e8_per_million,output_e8_per_million,cache_read_e8_per_million,cache_write_e8_per_million,cache_write_5m_e8_per_million,cache_write_1h_e8_per_million,image_input_e8_per_million,image_output_e8_per_million,priority_input_e8_per_million,priority_output_e8_per_million,priority_cache_read_e8_per_million,priority_cache_write_e8_per_million,fast_multiplier_bps,flex_multiplier_bps,max_reasoning_effort_multiplier_bps";
async function activeSnapshot(env: Pick<Env,"DB">): Promise<PricingSnapshot> {
  const rows = await env.DB.prepare("SELECT v.version_id,v.digest,v.max_reservation_e8_usd FROM pricing_active_version a JOIN pricing_versions v ON v.version_id=a.version_id").all<VersionRow>();
  if (rows.results.length !== 1) throw new PricingUnavailableError("missing or ambiguous active pricing version");
  const version = rows.results[0]; const rules = await env.DB.prepare(`SELECT ${ruleColumns} FROM pricing_rules WHERE version_id=?`).bind(version.version_id).all<PricingRule>();
  return validatePricingSnapshot(version.version_id,version.max_reservation_e8_usd,version.digest,rules.results);
}
export async function validateActivePricing(env: Pick<Env,"DB">): Promise<PricingSnapshot> { return activeSnapshot(env); }
export async function lookupAdmissionPriceCard(env: Pick<Env,"DB"|"CONFIG_CACHE">, model: string): Promise<AdmittedPriceCard> {
  const normalized = normalizePricingModel(model); if (!normalized) throw new PricingUnavailableError("invalid model");
  const snapshot = await activeSnapshot(env); const card = { version_id:snapshot.version_id,digest:snapshot.digest,max_reservation_e8_usd:snapshot.max_reservation_e8_usd,rule:selectRule(normalized,snapshot.rules) };
  try { await env.CONFIG_CACHE.put(`pricing:${card.version_id}:${normalized}`,JSON.stringify(card),{expirationTtl:300}); } catch { /* D1 authorized admission. */ }
  return card;
}
export async function loadAdmittedPriceCard(env: Pick<Env,"DB">, versionIDValue: string, versionDigest: string, pricingModel: string, modelPattern?: string, matchKind?: "exact" | "family"): Promise<AdmittedPriceCard> {
  const version = await env.DB.prepare("SELECT version_id,digest,max_reservation_e8_usd FROM pricing_versions WHERE version_id=? AND digest=?").bind(versionIDValue,versionDigest).first<VersionRow>();
  if (!version) throw new PricingUnavailableError("admitted price card missing");
  const rules = await env.DB.prepare(`SELECT ${ruleColumns} FROM pricing_rules WHERE version_id=?`).bind(versionIDValue).all<PricingRule>();
  const snapshot = await validatePricingSnapshot(version.version_id,version.max_reservation_e8_usd,version.digest,rules.results);
  const pattern = modelPattern ?? pricingModel;
  const normalized = normalizePricingModel(pricingModel); if (!normalized) throw new PricingUnavailableError("invalid admitted pricing model");
  const selected = selectRule(normalized,snapshot.rules);
  if (selected.model_pattern !== pattern || (matchKind !== undefined && selected.match_kind !== matchKind)) throw new PricingUnavailableError("admitted pricing rule mismatch");
  return {version_id:snapshot.version_id,digest:snapshot.digest,max_reservation_e8_usd:snapshot.max_reservation_e8_usd,rule:selected};
}
function price(value: string): bigint { return BigInt(value); }
function usage(value: unknown): bigint {
  if (!unsigned(value,MAX_USAGE_DIGITS)) throw new PricingUnavailableError("invalid usage"); return BigInt(value);
}
function halfUp(n: bigint, d: bigint): bigint { return n / d + ((n % d) * 2n >= d ? 1n : 0n); }
function nonzero(current: bigint, fallback: bigint): bigint { return current > 0n ? current : fallback; }
function tokenPart(tokens: bigint, unit: bigint, index: number, multiplier: bigint): Part { return { numerator: tokens * unit * multiplier, denominator: 1_000_000n * 10_000n * 10_000n * 10_000n, index }; }
function allocate(parts: Part[]): [bigint[],bigint] {
  const denominator = parts[0]?.denominator; if (!denominator || parts.some((part) => part.denominator !== denominator || part.numerator < 0n)) throw new PricingUnavailableError("invalid charge parts");
  const total = halfUp(parts.reduce((sum,part) => sum + part.numerator,0n),denominator); const values=parts.map((part) => part.numerator/denominator); const remainders=parts.map((part) => part.numerator%denominator);
  let delta=total-values.reduce((sum,value)=>sum+value,0n); if(delta<0n||delta>BigInt(parts.length))throw new PricingUnavailableError("invalid component allocation");
  const indices=parts.map((_,index)=>index).sort((a,b)=>remainders[a]===remainders[b]?parts[a].index-parts[b].index:remainders[a]>remainders[b]?-1:1);
  for(let i=0n;i<delta;i+=1n) values[indices[Number(i)]] += 1n; return [values,total];
}
function cacheCreation(total: bigint, five: bigint, hour: bigint): [bigint,bigint] {
  if (total === 0n) { if (five !== 0n || hour !== 0n) throw new PricingUnavailableError("cache detail without total"); return [0n,0n]; }
  if (five + hour <= total) return [five,hour];
  const detail=five+hour; if(detail===0n) throw new PricingUnavailableError("invalid cache detail"); const allocated=halfUp(total*five,detail); return allocated >= total ? [total,0n] : [allocated,total-allocated];
}
export function calculateAdmittedE8Charge(card: AdmittedPriceCard, value: UsageForPricing): E8ChargeBreakdown {
  if (!validRule(card.rule) || !positive(card.max_reservation_e8_usd,18) || !digest(card.digest) || !versionID(card.version_id) || card.rule.version_id !== card.version_id) throw new PricingUnavailableError("invalid admitted price card");
  const input=usage(value.input_tokens), imageInput=usage(value.image_input_tokens), output=usage(value.output_tokens), imageOutput=usage(value.image_output_tokens), creation=usage(value.cache_creation_tokens), creation5=usage(value.cache_creation_5m_tokens), creation1=usage(value.cache_creation_1h_tokens), read=usage(value.cache_read_tokens);
  if(imageInput>input||imageOutput>output) throw new PricingUnavailableError("impossible image usage subset");
  const rate=usage(value.rate_multiplier_bps); if(rate===0n || rate>99_999_999n) throw new PricingUnavailableError("invalid rate multiplier");
  let inputPrice=price(card.rule.input_e8_per_million),outputPrice=price(card.rule.output_e8_per_million),readPrice=price(card.rule.cache_read_e8_per_million),writePrice=price(card.rule.cache_write_e8_per_million),tierMultiplier=10_000n;
  const tier=value.service_tier.trim().toLowerCase();
  if(tier==="priority"||tier==="fast") { const fast=price(card.rule.fast_multiplier_bps); if(fast>0n)tierMultiplier=fast; else if(price(card.rule.priority_input_e8_per_million)>0n||price(card.rule.priority_output_e8_per_million)>0n||price(card.rule.priority_cache_read_e8_per_million)>0n||price(card.rule.priority_cache_write_e8_per_million)>0n){ inputPrice=nonzero(price(card.rule.priority_input_e8_per_million),inputPrice);outputPrice=nonzero(price(card.rule.priority_output_e8_per_million),outputPrice);readPrice=nonzero(price(card.rule.priority_cache_read_e8_per_million),readPrice);writePrice=nonzero(price(card.rule.priority_cache_write_e8_per_million),writePrice); }else tierMultiplier=20_000n; }
  if(tier==="flex") tierMultiplier=price(card.rule.flex_multiplier_bps)>0n?price(card.rule.flex_multiplier_bps):5_000n;
  const multiplier=tierMultiplier*rate*(value.reasoning_effort.trim().toLowerCase()==="max"?price(card.rule.max_reasoning_effort_multiplier_bps):10_000n);
  const [five,hour]=cacheCreation(creation,creation5,creation1); const write5=price(card.rule.cache_write_5m_e8_per_million),write1=price(card.rule.cache_write_1h_e8_per_million);
  const parts: Part[]=[]; const add=(tokens:bigint,unit:bigint) => parts.push(tokenPart(tokens,unit,parts.length,multiplier));
  add(input-imageInput,inputPrice); add(imageInput,nonzero(price(card.rule.image_input_e8_per_million),inputPrice)); add(output-imageOutput,outputPrice); add(imageOutput,nonzero(price(card.rule.image_output_e8_per_million),outputPrice));
  if(write5>0n||write1>0n) { add(five===0n&&hour===0n&&creation>0n?creation:five,write5); add(five===0n&&hour===0n&&creation>0n?0n:hour,write1); } else { add(creation,writePrice); add(0n,writePrice); }
  add(read,readPrice); const [values,total]=allocate(parts);
  return {input_e8_usd:values[0].toString(),image_input_e8_usd:values[1].toString(),output_e8_usd:values[2].toString(),image_output_e8_usd:values[3].toString(),cache_write_e8_usd:(values[4]+values[5]).toString(),cache_read_e8_usd:values[6].toString(),total_e8_usd:total.toString(),exceeds_reservation_cap:total>BigInt(card.max_reservation_e8_usd)};
}
// Backward-compatible narrow helper retained for callers that only have the
// legacy three counters. New settlement uses calculateAdmittedE8Charge.
export function calculateUsageChargeE8USD(card: AdmittedPriceCard,input:string,output:string,cacheRead:string): string {
  for (const value of [input,output,cacheRead]) if (!unsigned(value,MAX_USAGE_DIGITS)) throw new PricingUnavailableError("invalid usage");
  const numerator=BigInt(input)*price(card.rule.input_e8_per_million)+BigInt(output)*price(card.rule.output_e8_per_million)+BigInt(cacheRead)*price(card.rule.cache_read_e8_per_million);
  return halfUp(numerator,1_000_000n).toString();
}

File: backend/internal/cloudflarebridge/pricing.go
package cloudflarebridge

import (
	"errors"
	"math/big"
	"regexp"
	"sort"
	"strings"
)

type AdmittedPriceCard struct {
	VersionID           string        `json:"version_id"`
	Digest              string        `json:"digest"`
	MaxReservationE8USD string        `json:"max_reservation_e8_usd"`
	Rule                E8PricingRule `json:"rule"`
}

type E8PricingRule struct {
	VersionID                       string `json:"version_id"`
	ModelPattern                    string `json:"model_pattern"`
	MatchKind                       string `json:"match_kind"`
	InputE8PerMillion               string `json:"input_e8_per_million"`
	OutputE8PerMillion              string `json:"output_e8_per_million"`
	CacheReadE8PerMillion           string `json:"cache_read_e8_per_million"`
	CacheWriteE8PerMillion          string `json:"cache_write_e8_per_million"`
	CacheWrite5mE8PerMillion        string `json:"cache_write_5m_e8_per_million"`
	CacheWrite1hE8PerMillion        string `json:"cache_write_1h_e8_per_million"`
	ImageInputE8PerMillion          string `json:"image_input_e8_per_million"`
	ImageOutputE8PerMillion         string `json:"image_output_e8_per_million"`
	PriorityInputE8PerMillion       string `json:"priority_input_e8_per_million"`
	PriorityOutputE8PerMillion      string `json:"priority_output_e8_per_million"`
	PriorityCacheReadE8PerMillion   string `json:"priority_cache_read_e8_per_million"`
	PriorityCacheWriteE8PerMillion  string `json:"priority_cache_write_e8_per_million"`
	FastMultiplierBPS               string `json:"fast_multiplier_bps"`
	FlexMultiplierBPS               string `json:"flex_multiplier_bps"`
	MaxReasoningEffortMultiplierBPS string `json:"max_reasoning_effort_multiplier_bps"`
}

type E8Usage struct {
	InputTokens           int64
	ImageInputTokens      int64
	OutputTokens          int64
	ImageOutputTokens     int64
	CacheCreationTokens   int64
	CacheCreation5mTokens int64
	CacheCreation1hTokens int64
	CacheReadTokens       int64
	ServiceTier           string
	ReasoningEffort       string
	RateMultiplierBPS     string
}

type E8ChargeBreakdown struct {
	InputE8USD            string
	ImageInputE8USD       string
	OutputE8USD           string
	ImageOutputE8USD      string
	CacheWriteE8USD       string
	CacheReadE8USD        string
	TotalE8USD            string
	VersionID             string
	Digest                string
	ReservationCapE8USD   string
	ExceedsReservationCap bool
}

var (
	ErrInvalidAdmittedPriceCard = errors.New("invalid admitted price card")
	ErrInvalidE8Usage           = errors.New("invalid e8 usage")
	unsignedE8                  = regexp.MustCompile(`^(0|[1-9][0-9]*)$`)
	positiveE8                  = regexp.MustCompile(`^[1-9][0-9]*$`)
	digestE8                    = regexp.MustCompile(`^[0-9a-f]{64}$`)
	versionIDE8                 = regexp.MustCompile(`^[a-z0-9._:-]{1,128}$`)
	pricingModelE8              = regexp.MustCompile(`^[a-z0-9._:/-]+$`)
	priceDenominator            = big.NewInt(1_000_000)
	bpsDenominator              = big.NewInt(10_000)
)

func parseE8(value string, positive bool) (*big.Int, error) {
	if len(value) < 1 || len(value) > 40 ||
		(positive && !positiveE8.MatchString(value)) || (!positive && !unsignedE8.MatchString(value)) {
		return nil, ErrInvalidAdmittedPriceCard
	}
	v, ok := new(big.Int).SetString(value, 10)
	if !ok {
		return nil, ErrInvalidAdmittedPriceCard
	}
	return v, nil
}

func parseBPS(value string, positive bool) (*big.Int, error) {
	if len(value) > 8 {
		return nil, ErrInvalidAdmittedPriceCard
	}
	return parseE8(value, positive)
}

func normalizePricingModel(value string) (string, bool) {
	if len(value) < 1 || len(value) > 256 {
		return "", false
	}
	normalized := strings.ToLower(strings.TrimSpace(value))
	if normalized == "" || !pricingModelE8.MatchString(normalized) {
		return "", false
	}
	if strings.HasPrefix(normalized, "claude-") {
		normalized = strings.ReplaceAll(normalized, ".", "-")
	}
	return normalized, true
}

func ValidateAdmittedPriceCard(card AdmittedPriceCard) error {
	if !versionIDE8.MatchString(card.VersionID) || !digestE8.MatchString(card.Digest) ||
		card.Rule.VersionID != card.VersionID || card.Rule.MatchKind != "exact" && card.Rule.MatchKind != "family" {
		return ErrInvalidAdmittedPriceCard
	}
	if _, err := parseE8(card.MaxReservationE8USD, true); err != nil {
		return err
	}
	pattern := card.Rule.ModelPattern
	if len(pattern) < 1 || len(pattern) > 257 {
		return ErrInvalidAdmittedPriceCard
	}
	if card.Rule.MatchKind == "family" {
		if !strings.HasSuffix(pattern, "*") || strings.Contains(pattern[:len(pattern)-1], "*") {
			return ErrInvalidAdmittedPriceCard
		}
		base, ok := normalizePricingModel(pattern[:len(pattern)-1])
		if !ok || base+"*" != pattern {
			return ErrInvalidAdmittedPriceCard
		}
	} else {
		base, ok := normalizePricingModel(pattern)
		if !ok || base != pattern || strings.Contains(pattern, "*") {
			return ErrInvalidAdmittedPriceCard
		}
	}
	for _, value := range []string{card.Rule.InputE8PerMillion, card.Rule.OutputE8PerMillion, card.Rule.CacheReadE8PerMillion, card.Rule.CacheWriteE8PerMillion, card.Rule.CacheWrite5mE8PerMillion, card.Rule.CacheWrite1hE8PerMillion, card.Rule.ImageInputE8PerMillion, card.Rule.ImageOutputE8PerMillion, card.Rule.PriorityInputE8PerMillion, card.Rule.PriorityOutputE8PerMillion, card.Rule.PriorityCacheReadE8PerMillion, card.Rule.PriorityCacheWriteE8PerMillion, card.Rule.FastMultiplierBPS, card.Rule.FlexMultiplierBPS} {
		if _, err := parseE8(value, false); err != nil {
			return err
		}
	}
	if _, err := parseBPS(card.Rule.FastMultiplierBPS, false); err != nil {
		return err
	}
	if _, err := parseBPS(card.Rule.FlexMultiplierBPS, false); err != nil {
		return err
	}
	_, err := parseBPS(card.Rule.MaxReasoningEffortMultiplierBPS, true)
	return err
}

func ValidateAdmittedPriceCardForModel(card AdmittedPriceCard, model string) error {
	if err := ValidateAdmittedPriceCard(card); err != nil {
		return err
	}
	normalized, ok := normalizePricingModel(model)
	if !ok {
		return ErrInvalidAdmittedPriceCard
	}
	if card.Rule.MatchKind == "exact" && card.Rule.ModelPattern != normalized {
		return ErrInvalidAdmittedPriceCard
	}
	if card.Rule.MatchKind == "family" && !strings.HasPrefix(normalized, strings.TrimSuffix(card.Rule.ModelPattern, "*")) {
		return ErrInvalidAdmittedPriceCard
	}
	return nil
}

type exactPart struct {
	numerator   *big.Int
	denominator *big.Int
	index       int
}

func tokenPart(tokens int64, price *big.Int, index int) (exactPart, error) {
	if tokens < 0 {
		return exactPart{}, ErrInvalidE8Usage
	}
	return exactPart{numerator: new(big.Int).Mul(big.NewInt(tokens), price), denominator: new(big.Int).Set(priceDenominator), index: index}, nil
}

func multiply(part exactPart, bps *big.Int) exactPart {
	part.numerator.Mul(part.numerator, bps)
	part.denominator.Mul(part.denominator, bpsDenominator)
	return part
}

func normalizeCacheCreation(tokens E8Usage) (int64, int64, error) {
	if tokens.CacheCreationTokens < 0 || tokens.CacheCreation5mTokens < 0 || tokens.CacheCreation1hTokens < 0 {
		return 0, 0, ErrInvalidE8Usage
	}
	if tokens.CacheCreationTokens <= 0 || (tokens.CacheCreation5mTokens <= tokens.CacheCreationTokens && tokens.CacheCreation1hTokens <= tokens.CacheCreationTokens-tokens.CacheCreation5mTokens) {
		return tokens.CacheCreation5mTokens, tokens.CacheCreation1hTokens, nil
	}
	detail := new(big.Int).Add(big.NewInt(tokens.CacheCreation5mTokens), big.NewInt(tokens.CacheCreation1hTokens))
	if detail.Sign() <= 0 {
		return 0, 0, ErrInvalidE8Usage
	}
	// Traditional code uses math.Round for this ratio; for positive integer
	// tokens this is the same deterministic half-up allocation.
	n := new(big.Int).Mul(big.NewInt(tokens.CacheCreationTokens), big.NewInt(tokens.CacheCreation5mTokens))
	q, r := new(big.Int), new(big.Int)
	q.QuoRem(n, detail, r)
	if new(big.Int).Lsh(r, 1).Cmp(detail) >= 0 {
		q.Add(q, big.NewInt(1))
	}
	if q.Int64() >= tokens.CacheCreationTokens {
		return tokens.CacheCreationTokens, 0, nil
	}
	return q.Int64(), tokens.CacheCreationTokens - q.Int64(), nil
}

func nonzero(value *big.Int, fallback *big.Int) *big.Int {
	if value.Sign() > 0 {
		return value
	}
	return fallback
}

func allocate(parts []exactPart) ([]*big.Int, *big.Int, error) {
	if len(parts) == 0 || parts[0].denominator.Sign() <= 0 {
		return nil, nil, ErrInvalidE8Usage
	}
	den := new(big.Int).Set(parts[0].denominator)
	totalNumerator := big.NewInt(0)
	for _, part := range parts {
		if part.denominator.Cmp(den) != 0 || part.numerator.Sign() < 0 {
			return nil, nil, ErrInvalidE8Usage
		}
		totalNumerator.Add(totalNumerator, part.numerator)
	}
	total, remainder := new(big.Int), new(big.Int)
	total.QuoRem(totalNumerator, den, remainder)
	if new(big.Int).Lsh(remainder, 1).Cmp(den) >= 0 {
		total.Add(total, big.NewInt(1))
	}
	values, remainders := make([]*big.Int, len(parts)), make([]*big.Int, len(parts))
	sum := big.NewInt(0)
	for i, part := range parts {
		values[i], remainders[i] = new(big.Int), new(big.Int)
		values[i].QuoRem(part.numerator, den, remainders[i])
		sum.Add(sum, values[i])
	}
	order := make([]int, len(parts))
	for i := range order {
		order[i] = i
	}
	sort.SliceStable(order, func(i, j int) bool {
		c := remainders[order[i]].Cmp(remainders[order[j]])
		return c > 0 || (c == 0 && parts[order[i]].index < parts[order[j]].index)
	})
	delta := new(big.Int).Sub(total, sum)
	if !delta.IsInt64() || delta.Sign() < 0 || delta.Cmp(big.NewInt(int64(len(parts)))) > 0 {
		return nil, nil, ErrInvalidE8Usage
	}
	for i := int64(0); i < delta.Int64(); i++ {
		values[order[i]].Add(values[order[i]], big.NewInt(1))
	}
	return values, total, nil
}

func CalculateAdmittedE8Charge(card AdmittedPriceCard, usage E8Usage) (*E8ChargeBreakdown, error) {
	if err := ValidateAdmittedPriceCard(card); err != nil {
		return nil, err
	}
	for _, value := range []int64{usage.InputTokens, usage.ImageInputTokens, usage.OutputTokens, usage.ImageOutputTokens, usage.CacheReadTokens} {
		if value < 0 {
			return nil, ErrInvalidE8Usage
		}
	}
	p := func(value string) *big.Int { n, _ := parseE8(value, false); return n }
	input, output, read, write := p(card.Rule.InputE8PerMillion), p(card.Rule.OutputE8PerMillion), p(card.Rule.CacheReadE8PerMillion), p(card.Rule.CacheWriteE8PerMillion)
	tier := strings.ToLower(strings.TrimSpace(usage.ServiceTier))
	multiplier := big.NewInt(10_000)
	if tier == "fast" || tier == "priority" {
		fast := p(card.Rule.FastMultiplierBPS)
		if fast.Sign() > 0 {
			multiplier = fast
		} else if p(card.Rule.PriorityInputE8PerMillion).Sign() > 0 || p(card.Rule.PriorityOutputE8PerMillion).Sign() > 0 || p(card.Rule.PriorityCacheReadE8PerMillion).Sign() > 0 || p(card.Rule.PriorityCacheWriteE8PerMillion).Sign() > 0 {
			input, output, read, write = nonzero(p(card.Rule.PriorityInputE8PerMillion), input), nonzero(p(card.Rule.PriorityOutputE8PerMillion), output), nonzero(p(card.Rule.PriorityCacheReadE8PerMillion), read), nonzero(p(card.Rule.PriorityCacheWriteE8PerMillion), write)
		} else {
			multiplier = big.NewInt(20_000)
		}
	} else if tier == "flex" {
		if configured := p(card.Rule.FlexMultiplierBPS); configured.Sign() > 0 {
			multiplier = configured
		} else {
			multiplier = big.NewInt(5_000)
		}
	}
	imageIn := usage.ImageInputTokens
	if imageIn > usage.InputTokens {
		imageIn = usage.InputTokens
	}
	imageOut := usage.ImageOutputTokens
	if imageOut > usage.OutputTokens {
		imageOut = usage.OutputTokens
	}
	parts := make([]exactPart, 0, 7)
	appendPart := func(tokens int64, price *big.Int) error {
		part, err := tokenPart(tokens, price, len(parts))
		if err == nil {
			parts = append(parts, multiply(part, multiplier))
		}
		return err
	}
	if err := appendPart(usage.InputTokens-imageIn, input); err != nil {
		return nil, err
	}
	if err := appendPart(imageIn, nonzero(p(card.Rule.ImageInputE8PerMillion), input)); err != nil {
		return nil, err
	}
	if err := appendPart(usage.OutputTokens-imageOut, output); err != nil {
		return nil, err
	}
	if err := appendPart(imageOut, nonzero(p(card.Rule.ImageOutputE8PerMillion), output)); err != nil {
		return nil, err
	}
	five, hour, err := normalizeCacheCreation(usage)
	if err != nil {
		return nil, err
	}
	fivePrice, hourPrice := p(card.Rule.CacheWrite5mE8PerMillion), p(card.Rule.CacheWrite1hE8PerMillion)
	if fivePrice.Sign() > 0 || hourPrice.Sign() > 0 {
		if five == 0 && hour == 0 && usage.CacheCreationTokens > 0 {
			five = usage.CacheCreationTokens
		}
		if err := appendPart(five, fivePrice); err != nil {
			return nil, err
		}
		if err := appendPart(hour, hourPrice); err != nil {
			return nil, err
		}
	} else {
		if err := appendPart(usage.CacheCreationTokens, write); err != nil {
			return nil, err
		}
		if err := appendPart(0, write); err != nil {
			return nil, err
		}
	}
	if err := appendPart(usage.CacheReadTokens, read); err != nil {
		return nil, err
	}
	rate := usage.RateMultiplierBPS
	if rate == "" {
		rate = "10000"
	}
	rateBPS, err := parseBPS(rate, true)
	if err != nil {
		return nil, ErrInvalidE8Usage
	}
	for i := range parts {
		parts[i] = multiply(parts[i], rateBPS)
		if strings.EqualFold(strings.TrimSpace(usage.ReasoningEffort), "max") {
			max, _ := parseE8(card.Rule.MaxReasoningEffortMultiplierBPS, true)
			parts[i] = multiply(parts[i], max)
		}
	}
	values, total, err := allocate(parts)
	if err != nil {
		return nil, err
	}
	cap, _ := parseE8(card.MaxReservationE8USD, true)
	return &E8ChargeBreakdown{InputE8USD: values[0].String(), ImageInputE8USD: values[1].String(), OutputE8USD: values[2].String(), ImageOutputE8USD: values[3].String(), CacheWriteE8USD: new(big.Int).Add(values[4], values[5]).String(), CacheReadE8USD: values[len(values)-1].String(), TotalE8USD: total.String(), VersionID: card.VersionID, Digest: card.Digest, ReservationCapE8USD: cap.String(), ExceedsReservationCap: total.Cmp(cap) > 0}, nil
}
